### PR TITLE
sd-dhcp6-client: allow to build large packet

### LINKED ISSUE #24851: networkd DHCPv6 client fails to process large DHCPv6 packets
### systemd version the issue has been seen with

251

### Used distribution

Debian

### Linux kernel version used

5.19.0-2-cloud-amd64

### CPU architectures issue was seen on

x86_64

### Component

systemd-networkd

### Expected behaviour you didn't see

Configuration of IPv6 address and route state via DHCPv6 on a network interface in response to a DHCPv6 Advertise packet containing a large number of options.

### Unexpected behaviour you saw

Systemd-networkd fails to complete the DHCPv6 negotiation and never configures addresses or routes on the interface.

The problem _may_ be related to the size of the packet payload, as it's been observed that the IP payload size is >512 bytes in failure cases, while payload sizes <512 bytes seem to work.

### Steps to reproduce the problem

The problem has been observed in an Amazon Web Services VPC network environment in which a large number (approximately >15) number of IPv6 Delegated Prefixes have been associated with a network interface. In order to reproduce this, you'll need to run a fairly large instance (e.g. c6i.16xlarge), and there's a cost associated with this.

Create a network interface with 30 delegated prefixes associated with it, e.g.

```plain
$ aws ec2 create-network-interface \
--subnet-id subnet-047cfed18eEXAMPLE \
--ipv6-address-count 1 \
--ipv6-prefix-count 30
```

Launch an instance the current Debian unstable AMI. You can obtain the AMI ID with a command similar to the following (adjusted for region, etc):
```plain
$ aws ec2 describe-images --owner 903794441882 --region us-west-2 --output json \
  --query "Images[?Architecture=='x86_64'] | [?starts_with(Name, 'debian-sid-')] | max_by([], &CreationDate)"
```
Install a .network file for the previously created ENI on the instance (adjust the MAC address to match the one from your ENI):

```plain
[Match]
Driver=ena
MACAddress=06:93:f5:28:17:39

[Link]
MTUBytes=9001

[Network]
DHCP=ipv6
LLMNR=no

[DHCPv6]
UseHostname=no
UseDNS=no
UseNTP=no

[Route]
Gateway=_ipv6ra

[IPv6AcceptRA]
UseDomains=no
```

Attach the interface to the instance, e.g.

```
$ aws ec2 attach-network-interface --output text --device-index 1 \
  --instance-id $iid --network-interface-id $eni_id --network-card-index 0
```

The same steps, but with a smaller number of prefixes assigned to the network interface (say 3 instead of 30) yields a successful DHCPv6 negotiation, and addresses and routes are configured as expected.

### Additional program output to the terminal or log subsystem illustrating the issue

```sh
This problem has also been observed with systemd 250.8 on the Amazon Linux 2022.

With debug output enabled for systemd-networkd, we see the following, which indicates that we've sent a Solicit message and processed an Advertise response, but we never seem to actually send a Request message.


$ journalctl -u systemd-networkd.service -b | grep ens6 | grep DHCPv6 | head -n30
Sep 28 17:12:21 ip-10-0-1-71 systemd-networkd[1095]: ens6: Requested configuring of the DHCPv6 client.
Sep 28 17:12:21 ip-10-0-1-71 systemd-networkd[1095]: ens6: DHCPv6 client is configured.
Sep 28 17:12:22 ip-10-0-1-71 systemd-networkd[1095]: ens6: link_check_ready(): DHCPv4, DHCPv6, DHCP-PD or IPv4LL is enabled but no dynamic address is assigned yet.
Sep 28 17:12:22 ip-10-0-1-71 systemd-networkd[1095]: ens6: link_check_ready(): DHCPv4, DHCPv6, DHCP-PD or IPv4LL is enabled but no dynamic address is assigned yet.
Sep 28 17:12:24 ip-10-0-1-71 systemd-networkd[1095]: ens6: DHCPv6 client: Starting in Solicit mode
Sep 28 17:12:24 ip-10-0-1-71 systemd-networkd[1095]: ens6: DHCPv6 client: State changed: stopped -> solicitation
Sep 28 17:12:24 ip-10-0-1-71 systemd-networkd[1095]: ens6: Acquiring DHCPv6 lease on NDisc request
Sep 28 17:12:24 ip-10-0-1-71 systemd-networkd[1095]: ens6: link_check_ready(): DHCPv4, DHCPv6, DHCP-PD or IPv4LL is enabled but no dynamic address is assigned yet.
Sep 28 17:12:24 ip-10-0-1-71 systemd-networkd[1095]: ens6: DHCPv6 client: Sent Solicit
Sep 28 17:12:24 ip-10-0-1-71 systemd-networkd[1095]: ens6: DHCPv6 client: Next retransmission in 1s
Sep 28 17:12:24 ip-10-0-1-71 systemd-networkd[1095]: ens6: link_check_ready(): DHCPv4, DHCPv6, DHCP-PD or IPv4LL is enabled but no dynamic address is assigned yet.
Sep 28 17:12:24 ip-10-0-1-71 systemd-networkd[1095]: ens6: DHCPv6 client: Processed Advertise message
Sep 28 17:12:24 ip-10-0-1-71 systemd-networkd[1095]: ens6: DHCPv6 client: State changed: solicitation -> request
Sep 28 17:12:24 ip-10-0-1-71 systemd-networkd[1095]: ens6: DHCPv6 client: Next retransmission in 912ms
Sep 28 17:12:25 ip-10-0-1-71 systemd-networkd[1095]: ens6: DHCPv6 client: Next retransmission in 1s
Sep 28 17:12:26 ip-10-0-1-71 systemd-networkd[1095]: ens6: DHCPv6 client: Next retransmission in 3s
Sep 28 17:12:30 ip-10-0-1-71 systemd-networkd[1095]: ens6: DHCPv6 client: Next retransmission in 6s
Sep 28 17:12:32 ip-10-0-1-71 systemd-networkd[1095]: ens6: Acquiring DHCPv6 lease on NDisc request
Sep 28 17:12:32 ip-10-0-1-71 systemd-networkd[1095]: ens6: link_check_ready(): DHCPv4, DHCPv6, DHCP-PD or IPv4LL is enabled but no dynamic address is assigned yet.
Sep 28 17:12:32 ip-10-0-1-71 systemd-networkd[1095]: ens6: link_check_ready(): DHCPv4, DHCPv6, DHCP-PD or IPv4LL is enabled but no dynamic address is assigned yet.
Sep 28 17:12:34 ip-10-0-1-71 systemd-networkd[1095]: ens6: link_check_ready(): DHCPv4, DHCPv6, DHCP-PD or IPv4LL is enabled but no dynamic address is assigned yet.
```

### PR DESCRIPTION
**Pull request by yuwata**
Fixes #24851.

### CODE REVIEW

**Comment by poettering** [comment]
hmm, any chance you can separate the buflen/offset renaming thing into a separate patch? it's so hard to review if the renaming and the actual change of code is mangled into one

**Comment by nmeyerhans** [comment]
Looks good in my testing. The dhcpv6 client is able to complete a full negotiation cycle involving large numbers of IA_PD options and large packets.

**Comment by yuwata** [comment]
> hmm, any chance you can separate the buflen/offset renaming thing into a separate patch? it's so hard to review if the renaming and the actual change of code is mangled into one

The renaming `buflen` -> `offset` is crucial, as the stored values are changed. Previously, the size of the free are in the packet was taken, but now the offset to the free area is taken. I mentioned that change in the commit message.

Instead, now I split out many tiny changes/cleanups that are previously squashed to the main change. I hope that makes easier to review this PR.

PTAL!

**Comment by yuwata** [comment]
@nmeyerhans Thank you for testing the PR!

File: src/libsystemd-network/dhcp6-option.c [hunk 1]
```diff
@@ -322,21 +313,13 @@ static int option_append_pd_prefix(uint8_t **buf, size_t *buflen, const struct i
                 .address = prefix->address,
         };
```
Comment (keszybz) [inline src/libsystemd-network/dhcp6-option.c:318]: In fact, this could be simplified further by defining `p` here. Then the cast becomes unnecessary.

**Review by keszybz — APPROVED** [review]
LGTM.

Since this was already tested by @nmeyerhans, it looks good to go.

File: src/libsystemd-network/dhcp6-option.c (hunk 1 above)
Comment (keszybz) [reply src/libsystemd-network/dhcp6-option.c:318]: → https://github.com/systemd/systemd/pull/24973

**Pull request merged**